### PR TITLE
Fixing reverse index mask bug

### PR DESCRIPTION
**Pull request by avgAGB**

### CODE REVIEW

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java [hunk 1]
```diff
@@ -679,19 +680,37 @@ protected Value createBloomFilter(RawRecordContainer event, Multimap<String,Norm
...
             Text colf = new Text(column);
             Text colq = new Text(shardId);
             TextUtil.textAppend(colq, event.getDataType().outputName(), helper.getReplaceMalformedUTF8());
             
-            // Dont create index entries for empty values
-            if (!StringUtils.isEmpty(normalizedMaskedValue)) {
-                // Create a key for the masked field value with the masked visibility
-                Key k = this.createIndexKey(normalizedMaskedValue.getBytes(), colf, colq, maskedVisibility, event.getDate(), false);
+            // if this method was called with the intention to create reverse index keys, ensure the masked values are reversed.
+            if (isReverse) {
```
Comment (ivakegg) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: I think this whole section could be much simpler by switching the if statements:
```
if (!StringUtils.isEmpty(normalizedMaskedValue)) {
   if (isReverse) {
     normalizedMaskedValue = new StringBuilder().append(normalizedMaskedValue).reverse().toString();
   }
   Key k = this.createIndexKey(normalizedMaskedValue.getBytes(), colf, colq, maskedVisibility, event.getDate(0, false);
   BulkIngestKey bKey = BulkIngestKey(tableName, k);
   values.put(bkey, indexValue);
}
```

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java (hunk 1 above)
Comment (avgAGB) [reply warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: I'll adjust the ordering on the next commit.

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java [hunk 2]
```diff
@@ -679,21 +680,26 @@ protected Value createBloomFilter(RawRecordContainer event, Multimap<String,Norm
...
             Text colf = new Text(column);
             Text colq = new Text(shardId);
             TextUtil.textAppend(colq, event.getDataType().outputName(), helper.getReplaceMalformedUTF8());
             
-            // Dont create index entries for empty values
+            // if this method was called with the intention to create reverse index keys, ensure the masked values are reversed.
             if (!StringUtils.isEmpty(normalizedMaskedValue)) {
-                // Create a key for the masked field value with the masked visibility
+                if (isReverse) {
+                    normalizedMaskedValue = new StringBuilder().append(normalizedMaskedValue).reverse().toString();
```
Comment (jschmidt10) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: Can be simplified a little: `new StringBuilder(normalizedMaskedValue).reverse().toString()`

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java [hunk 3]
```diff
@@ -655,9 +655,10 @@ protected Value createBloomFilter(RawRecordContainer event, Multimap<String,Norm
      * @param shardId
      * @param tableName
      * @param indexValue
+     * @param isReverse
      */
     protected Multimap<BulkIngestKey,Value> createTermIndexColumn(RawRecordContainer event, String column, String fieldValue, byte[] visibility,
-                    byte[] maskedVisibility, MaskedFieldHelper maskedFieldHelper, byte[] shardId, Text tableName, Value indexValue) {
+                    byte[] maskedVisibility, MaskedFieldHelper maskedFieldHelper, byte[] shardId, Text tableName, Value indexValue, boolean isReverse) {
```
Comment (jschmidt10) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: Might not be worth the effort here but we do have a `Direction` enum. Unfortunately it's in datawave-ws-client. We could bump that guy to datawave-core and then use it here. Using booleans as flags for things that don't really have "truthy" value to them sorta requires you to look at the method definition every time to remind yourself what true vs false means. 

```
createTermIndexColumn(..., indexValue, false); // requires drilling into the method to make sure you used the right value
```
VS
```
createTermIndexColumn(..., indexValue, Direction.FORWARD); // is unambiguously creating the forward values
```
Thoughts? Bueller?

**Review by matthpeterson — CHANGES_REQUESTED** [review]
Looks good!  I'd like to see some tests that capture the bug behavior and show that it's now fixed.

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java (hunk 3 above)
Comment (avgAGB) [reply warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: I'll work it into the follow-on commits, that sounds good.

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java (hunk 2 above)
Comment (avgAGB) [reply warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: Alright, will adjust in the next commit.

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java [hunk 4]
```diff
@@ -679,21 +680,26 @@ protected Value createBloomFilter(RawRecordContainer event, Multimap<String,Norm
...
             
             Text colf = new Text(column);
             Text colq = new Text(shardId);
             TextUtil.textAppend(colq, event.getDataType().outputName(), helper.getReplaceMalformedUTF8());
             
-            // Dont create index entries for empty values
+            // if this method was called with the intention to create reverse index keys, ensure the masked values are reversed.
             if (!StringUtils.isEmpty(normalizedMaskedValue)) {
-                // Create a key for the masked field value with the masked visibility
+                if (direction.getValue().equals("reverse")) {
```
Comment (jschmidt10) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: The nice thing about enum constants is that the JVM only has 1 instance of each. You can do this: `direction == Direction.REVERSE` which is super fast.

File: warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java [hunk 5]
```diff
@@ -1,17 +1,176 @@
...
+            String rawRecord = new String(record.getRawData());
+            for (String value : rawRecord.split(";")) {
+                NormalizedContentInterface geo_nci = new NormalizedFieldAndValue("FIELD_NAME", value);
+                eventFields.put("FIELD_NAME", geo_nci);
+            }
+            return normalizeMap(eventFields);
+        }
+        
+        public String getNormalizedMaskedValue(final String key) {
+            return "FIELD_NAME";
```
Comment (jschmidt10) [inline warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java]: Might be worth making this `MASKED_VALUE` or something like that. Your unit test below asserts that "FIELD_NAME" reversed is returned. But that had me scratching my head wondering "Why are we reversing the FIELD_NAME?". It just so happens you pass in FIELD_NAME to the createTermIndexColumn() in your test too.

If you change this to MASKED_VALUE it'll make it more obvious that this is a dummy version of a masked value an different from the field name.

File: warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java [hunk 6]
```diff
@@ -1,17 +1,176 @@
...
+        byte[] shardId = new byte[] {50, 48, 48, 48, 48, 49, 48, 49, 95, 54, 57};
+        
+        Multimap<BulkIngestKey,Value> termIndex = handler.createTermIndexColumn(record, "FIELD_NAME", "FIELD_VALUE", visibility, null, null, shardId,
+                        handler.getShardIndexTableName(), new Value(uid.toByteArray()), Direction.FORWARD);
+        
+        assertTrue(termIndex.size() == 1);
+    }
+    
+    @Test
+    public void testCreateTermReverseIndex() {
```
Comment (jschmidt10) [inline warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java:152]: Nice test!

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java (hunk 4 above)
Comment (avgAGB) [reply warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandler.java]: Updated in lastest commit.

File: warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java (hunk 5 above)
Comment (avgAGB) [reply warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java]: Updated in lastest commit

File: warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java (hunk 6 above)
Comment (avgAGB) [reply warehouse/ingest-core/src/test/java/datawave/ingest/mapreduce/handler/shard/ShardedDataTypeHandlerTest.java:152]: Thanks! Going to iterate tests on top of this.

**Review by jschmidt10 — DISMISSED** [review]

**Review by matthpeterson — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]

**Pull request merged**